### ISSUE TITLE
HorizontalPodAutoscaler graduation requirements

### ISSUE THREAD

**Issue by bgrant0607** [kubernetes/kubernetes#15315]
Before HorizontalPodAutoscaler graduates from experimental:
* It needs to implement a status subresource
* It needs to adhere to API conventions. For instance, LastScaleTimestamp should be LastScaleTime (see naming section of conventions)

Before it goes to beta:
* Target field needs a more descriptive name

Found these things at a glance. Will do a real review later.

cc @fgrzadkowski @jszczepkowski

**Comment by brendandburns** [comment]
I don't see any of these as blockers to leaving experimental or going into Beta.

   * Adding a status is an additive change.
   * API conventions can be achieved via deprecating names and copying on the way into stable
   * as above wrt to Target.

**Comment by bgrant0607** [comment]
cc @davidopp

**Comment by bgrant0607** [comment]
Status is a breaking change.

**Comment by bgrant0607** [comment]
The problem with not obeying API conventions is basically the broken window theory. It encourages others to do the same, even if unintentionally.

**Comment by davidopp** [comment]
Why is adding Status a breaking change?

**Comment by jszczepkowski** [comment]
CC @derekwaynecarr 
We have some other substantial requirements for the API before going to *beta*:
* support for multiple targets,
* support for application-level/custom metrics,
* (maybe) support for aggregated metrics.
Especially support for application level metrics is important. The current autoscaling based only on CPU or memory seems to be not sufficient for customers.

**Comment by jszczepkowski** [comment]
Regarding API changes describe here by @bgrant0607: I totally agree they should be applied ASAP.

**Comment by bgrant0607** [comment]
@davidopp HorizontalPodAutoscaler has a status field already, but it is written by writing the whole object rather than a `/status` subresource. Adding the latter enables ignoring status writes on POST/PUT/PATCH to the former. But that is a breaking change.

**Comment by derekwaynecarr** [comment]
@brendandburns - the PR to change to do ```/status``` resource has been done.

https://github.com/kubernetes/kubernetes/pull/14568

It would be good to get it merged as I have been a rebase loop (which is fine since many of the rebase requirements were because i added /status to the other experimental apis.)

**Comment by derekwaynecarr** [comment]
@jszczepkowski 

>> support for multiple targets

Is this just a matter of doing the following:
1. make ```HorizontalPodAutoscalerSpec.Target``` a slice?
2. make ```HorizontalPodAutoscalerStatus.CurrentConsumption``` a slice?

>> support for application-level/custom metrics

Does this require an API change, or just ensuring our validation of target resource names allows for values like ```myapp.com/custom-resource".  I assume in the case of custom-metrics a third-party controller is updating the status to denote the CurrentConsumption for those custom metrics.  We would then need to ensure that we do the right thing in response in the controllers (i.e. ignoring / preserving custom metric values from status generation)?

**Comment by jszczepkowski** [comment]
> > support for multiple targets

> Is this just a matter of doing the following:
> 1. make HorizontalPodAutoscalerSpec.Target a slice?
> 2. make HorizontalPodAutoscalerStatus.CurrentConsumption a slice?

Yes. This is a small issue.

> support for application-level/custom metrics

> > Does this require an API change, or just ensuring our validation of target resource names allows for values like myapp.com/custom-resource. I assume in the case of custom-metrics a third-party controller is updating the status to denote the CurrentConsumption for those custom metrics. We would then need to ensure that we do the right thing in response in the controllers (i.e. ignoring / preserving custom metric values from status generation)?

We will need to change type of ``Resource`` in ``ResourceConsumption`` to be an arbitrary string which defines application metric (or a new alias to string). We may also need to somehow distinguish application metrics from embedded CPU/memory metrics implemented today. For me, it is bit hard to predict how it should look like before we agree on general design of application metrics in Kubernetes. Maybe embedded Kubernetes metrics  should be prefixed with "kubernetes.io/"?

**Comment by bgrant0607** [comment]
In Deployment, we enabled maxUnavailable and maxSurge to be expressed in both absolute and relative (n%) terms. Relative works much better in the presence of vertical scaling, either automatic or manual, and is fairly intuitive to reason about (e.g., 50% utilization). However, there are probably some cases where absolute is more natural.

**Comment by bgrant0607** [comment]
@jszczepkowski mentioned that https://github.com/kubernetes/heapster/issues/642 is important to resolve.

**Comment by jszczepkowski** [comment]
Currently, targets in autoscaler are absolute. Our plan is to support relative targets in future by introducing new, relative, metrics to heapster. So, we can have a string for metric, which will currently support two values: "kubernetes.io/cpu-usage" and "kubernetes.io/memory-usage". In future (version 1.2), we can extend this by other embedded metrics: "kubernetes.io/cpu-usage-relative" and "kubernetes.io/memory-usage-relative" and by application metrics.

@bgrant0607 @derekwaynecarr Is it OK for you?

**Comment by bgrant0607** [comment]
@jszczepkowski Is there a proposal for those relative metrics somewhere? It seems like a hack, but maybe I don't understand the motivation.

**Comment by derekwaynecarr** [comment]
@jszczepkowski 

>> api.ResourceName to an arbitrary string...

It should already be able to be a relative string, maybe I am missing something non-obvious, but I don't see how this is different than anywhere else we handle compute resources with respect to allowing custom namespace prefixing to denote non standard resources.

**Comment by bgrant0607** [comment]
Prefixing built-in resources with kubernetes.io should not be necessary:
https://github.com/kubernetes/kubernetes/blob/master/docs/design/resources.md#resource-types

**Comment by jszczepkowski** [comment]
> Is there a proposal for those relative metrics somewhere? It seems like a hack, but maybe I don't understand the motivation.

Please see: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/horizontal-pod-autoscaler.md#relative-vs-absolute-metrics

**Comment by jszczepkowski** [comment]
> It should already be able to be an arbitrary string, maybe I am missing something non-obvious, but I don't see how this is different than anywhere else we handle compute resources with respect to allowing custom namespace prefixing to denote non standard resources.

...
 > Prefixing built-in resources with kubernetes.io should not be necessary ...

So, you proposition is to keep using api.ResourceName and to treat application metric as a Kubernetes Resource? I thought that Resource is something that can be requested/allocated. I'm not sure is it possible/reasonable to allocate a metric.

**Comment by bgrant0607** [comment]
Not a blocker for v1beta1:

minReplicas and maxReplicas should be optional, with minReplicas defaulting to 1 and maxReplicas defaulting to infinity (which is representable by nil); in Go, they'd both have to be pointers to int

**Comment by bgrant0607** [comment]
Sorry if this was already discussed -- point me at the thread, if so: Is a single set of target numbers desirable? Similar to how it's useful to have separate thresholds for heat and AC, it seems useful to have separate thresholds for scaling up and down. 

I'd also like to think about how we could provide default values for the thresholds based on the QoS class. Speaking of which, that raises the question of whether relative thresholds would be relative to the request or limit.

**Comment by bgrant0607** [comment]
targetConsumption would be a better name for a single absolute target. targetUtilization would be a better name for a single relative target.

Also, we generate user docs from types.go. Users shouldn't be expected to read the code. Therefore, all the API comments should be in terms of json field names rather than Go field names, and type-level comments should use plain English rather than Go type names. See #15350 for an example.

**Comment by bgrant0607** [comment]
With regards to relative metrics: Increasing the number of resource types and the metrics data volume does not seem preferable to just allowing utilization targets to be specified, which looks like the direction that node scaling is going, as well.

We should make the targetConsumption optional (needs omitempty) and specify that we'll add more ways to specify targets in the future, which would allow us to treat it as the sole member of a "oneof" group. In validation we can confirm that it is specified.

With regards to one threshold or two: I see some tolerance is built in. We can start with that, though I'd guess that such a small tolerance won't be sufficient. We should be able to increase it later, or perhaps autotune the tolerance.

I'd hope that we could autotune the hystersis time periods, so we shouldn't need to add them to the API.

**Comment by bgrant0607** [comment]
Regarding the rule-based approach mentioned here: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/horizontal-pod-autoscaler.md#next-steps

Rather than rules of the form "if average CPU consumption of pod is higher than 80% add two more replicas", I'd imagine we'd add the number of replicas required to decrease utilization per pod by the desired amount. This is similar to what a PID controller does -- the farther away from its goal, the bigger the change in output. For instance, if the target utilization were 50% and all the pods were at 100%, then the number of pods could be doubled. I could see a case for capping the amount of increase, but hopefully you get the idea.

**Comment by bgrant0607** [comment]
And why is ScaleRef a pointer if it's required?

And Resource shouldn't have omitempty, since it must be required, and it should be mergeable.

Also, what does CurrentConsumption mean? How does it relate to the proposed compute metrics API? The comment says average. Did you really mean `mean`? If so, why? Measured over what time period? Sampled when? Aggregated over all the pods how?
https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/compute-resource-metrics-api.md

Anyway, I'd prefer not to put usage data into configured resources, as explained here: https://github.com/kubernetes/kubernetes/blob/master/docs/design/resources.md#usage-data

I guess since the pod metrics API doesn't exist yet, you aren't using it? @vishh What's the status of the pod metrics API?

In addition to LastScaleTime, ObservedGeneration would be useful, for determining whether a change in the spec had been observed by the autoscaler.

Roll up of recommended changes:
```
type HorizontalPodAutoscalerSpec struct {
	ScaleRef SubresourceReference `json:"scaleRef"`
	// default to 1
	MinReplicas *int `json:"minReplicas,omitempty"`
	// default to no max
	MaxReplicas *int `json:"maxReplicas,omitempty"`
	TargetConsumptions []ResourceConsumption `json:"targetConsumptions" patchStrategy:"merge" patchMergeKey:"resourceName"`
}
type ResourceConsumption struct {
	ResourceName v1.ResourceName `json:"resourceName"`
	// only one of the following three should be specified
	Quantity *resource.Quantity `json:"quantity,omitempty"` // absolute
	requestPercentage *int `json:"requestPercentage,omitempty"` // utilization
}
type HorizontalPodAutoscalerStatus struct {
	ObservedGeneration *int64 `json:"observedGeneration,omitempty"`
	LastScaleTime *unversioned.Time `json:"lastScaleTime,omitempty"`
	CurrentReplicas int `json:"currentReplicas"`
	DesiredReplicas int `json:"desiredReplicas"`
}
```

**Comment by bgrant0607** [comment]
Actually, why did you choose "consumption" rather than "utilization"? I think the latter could work even for absolute quantities.

Minimal changes:
* in HorizontalPodAutoscalerSpec, rename target to targetUtilizations and make it an array
* in ResourceConsumption, resource needs to be required and quantity needs to be optional
* remove currentConsumption

**Comment by davidopp** [comment]
What is the difference between "mean" and "average" ?

**Comment by davidopp** [comment]
> We should make the targetConsumption optional

What would the target be in that case? Elsewhere you suggest we could infer from QoS class, but I'd argue that such inference is the more sophisticated feature, with manual-specification being the initial version of the feature. (You seem to think the opposite, i.e. defaulting based on QoS class should be the initial version, and then ways to manually specify should be the more sophisticated version, but I disagree -- among other things, we need some experience in production before we know what are good values for defaulting.)

**Comment by davidopp** [comment]
> remove currentConsumption

For debugging and understandability, I think it's useful to show the user what is the input going into the auto-scaling decision; IIUC that's what currentConsumption is.

**Comment by jszczepkowski** [comment]
> minReplicas and maxReplicas should be optional, with minReplicas defaulting to 1 and maxReplicas defaulting to infinity (which is representable by nil); in Go, they'd both have to be pointers to int

Defaulting minReplicas to 1 is fine. However, I'm not sure if defaulting maxReplicas to infinity is resonable. Please note that wrongly set utilization target (too small value), or an error in user app may cause huge grow of the number of instances and DOS the whole cluster.

**Comment by jszczepkowski** [comment]
> Sorry if this was already discussed -- point me at the thread, if so: Is a single set of target numbers desirable? Similar to how it's useful to have separate thresholds for heat and AC, it seems useful to have separate thresholds for scaling up and down.

Yes. In case of multiple metrics: we calculate the desired number of replicas independently for each of them and choose the biggest number.

**Comment by jszczepkowski** [comment]
> > remove currentConsumption

> For debugging and understandability, I think it's useful to show the user what is the input going into the auto-scaling decision; IIUC that's what currentConsumption is.

I agree with @davidopp: this value is extremely useful for debugging, kubectl get hpa prints it.

**Comment by jszczepkowski** [comment]
> And why is ScaleRef a pointer if it's required?

It shouldn't be. I'll change it.

> And Resource shouldn't have omitempty, since it must be required, and it should be mergeable.

I'll change it.

> Also, what does CurrentConsumption mean? How does it relate to the proposed compute metrics API? The comment says average. Did you really mean mean? If so, why? Measured over what time period? Sampled when? Aggregated over all the pods how?
https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/compute-resource-metrics-api.md

This is the arithmetic mean. I can rename it or update the comment.

> I guess since the pod metrics API doesn't exist yet, you aren't using it? @vishh What's the status of the pod metrics API?

I'm not using metrics API yet.

> In addition to LastScaleTime, ObservedGeneration would be useful, for determining whether a change in the spec had been observed by the autoscaler.

Do you mean here "the generation of HorizontalPodAutoscalerSpec"?

**Comment by jszczepkowski** [comment]
> targetConsumption would be a better name for a single absolute target. targetUtilization would be a better name for a single relative target.

> ```
type ResourceConsumption struct {
    ResourceName v1.ResourceName `json:"resourceName"`
    // only one of the following three should be specified
    Quantity *resource.Quantity `json:"quantity,omitempty"` // absolute
    requestPercentage *int `json:"requestPercentage,omitempty"` // utilization
}
```
We cannot have two different target fields for relative and absolute metrics. Whether the metrics is relative or absolute is the property of the metric. On HPA level we don't know it. The metric can be arbitrary defined by user.

**Comment by jszczepkowski** [comment]
> ```
type ResourceConsumption struct {
    ResourceName v1.ResourceName `json:"resourceName"`
...
}
```

@bgrant0607 As I said in https://github.com/kubernetes/kubernetes/issues/15315#issuecomment-147528733, I'm not sure if ``v1.ResourceName`` is a suitable type here. We want to make autoscaler more generic, it needs to scale on a base of **metric** (not **resource**).

So, therefore, I think we should introduce a new type:
```
type MetricName string
```
We should rename ``ResourceConsumption`` -> ``MetricConsumption``. For 1.1, we will support only two, embedded, absolute metrics:  "kubernetes.io/cpu-usage" and "kubernetes.io/memory-usage" (enforced by validation). For 1.2, additional embedded metrics (relative) and application level metrics will become available.

**Comment by jszczepkowski** [comment]
After consultation w/ @mwielgus: format "kubernetes.io:cpu-usage" is better than " "kubernetes.io/cpu-usage", as we will prevent future problems with interpreting "/" as a part of URL path in heapster.

**Comment by bgrant0607** [comment]
It sounds like the direction you want to go is a generic metric optimizer. In that case, neither "consumption" nor "utilization" is appropriate. There would just be targetMetrics, and currentMetrics, if we retain that field. If we were going to be rigorous, the metric `name` should then include details such as the time period for the metric, similar to alerting rules. However, that would probably be too complex for most users.

Please don't introduce a new namespacing syntax into the system. Can you give a concrete example of a heapster path you are concerned about? Why wouldn't you strip off the kubernetes.io/ prefix? Why do you even have a kubernetes.io/ prefix?

**Comment by bgrant0607** [comment]
BTW, what do you imagine the autoscaler should do in the case that multiple metric targets weren't simultaneously achievable?

**Comment by derekwaynecarr** [comment]
@jszczepkowski - I view a resource as anything that can be quantified or measured.

If an application has a mechanism to quantify an application metric (i.e. site qps), then that seems like a reasonable thing to treat as a custom prefixed resource (i.e. myapp.com/qps).  Not all of our ```api.ResourceName``` tokens denote traditional compute resources.  For example, quota has tokens for ```api.ResourceName``` objects that align to quantity of pod, service, etc.

**Comment by jszczepkowski** [comment]
@bgrant0607 

> Please don't introduce a new namespacing syntax into the system. Can you give a concrete example of a heapster path you are concerned about? Why wouldn't you strip off the kubernetes.io/ prefix? Why do you even have a kubernetes.io/ prefix?

There will be (at least) two types of heapster metrics: system metrics (e.g.: computed by Heapster for all pods) and application metircs (exposed by user apps). So, "kubernetes.io" prefix indicates we deal with system metrics.

> BTW, what do you imagine the autoscaler should do in the case that multiple metric targets weren't simultaneously achievable?

We choose the most pessimistic one (i.e.: the one that results in creation of the largest number of pods, https://github.com/kubernetes/kubernetes/issues/15315#issuecomment-147632531). It will be similar to how multiple metrics in autoscaling of nodes are currently implemented: https://github.com/kubernetes/kubernetes/blob/master/docs/admin/cluster-management.md#horizontal-auto-scaling-of-nodes-gce.

**Comment by jszczepkowski** [comment]
@derekwaynecarr 
> I view a resource as anything that can be quantified or measured.

> If an application has a mechanism to quantify an application metric (i.e. site qps), then that seems like a reasonable thing to treat as a custom prefixed resource (i.e. myapp.com/qps). Not all of our api.ResourceName tokens denote traditional compute resources. For example, quota has tokens for api.ResourceName objects that align to quantity of pod, service, etc.

Using ResourceName will require additional layer of translation from Resource to the corresponding heapster metric. I think it is simpler and better to just specify the heapster metric name.

**Comment by derekwaynecarr** [comment]
@jszczepkowski - isn't that mapping pretty much required when we couple with setting initial resources?  see https://github.com/kubernetes/kubernetes/pull/15476/files#diff-13e5957b77240f79aaedc3be6bf26d82R53

**Comment by jszczepkowski** [comment]
I've created PR #15549 for this. I suggest we move all new discussion there, as having the PR will make the discussion more concrete.

**Comment by jszczepkowski** [comment]
> isn't that mapping pretty much required when we couple with setting initial resources? see https://github.com/kubernetes/kubernetes/pull/15476/files#diff-13e5957b77240f79aaedc3be6bf26d82R53

Yes, but for IR we have only two absolute metrics: memory and CPU. For HPA we may much more metrics, e.g.: relative CPU, relative memory, application metrics.

E.g.: you may have a QPS resource, but it may refer to different application level metrics with different name.

E.g.2.: I imagine it hard to introduce yet another resource for relative CPU and memory.

**Comment by jszczepkowski** [comment]
@bgrant0607 
> Please don't introduce a new namespacing syntax into the system. Can you give a concrete example of a heapster path you are concerned about? Why wouldn't you strip off the kubernetes.io/ prefix?

Imagine, we have "cpu-usage" metric in Heapster's. If we want to query it by model API, we call:
``/api/v1/model/metrics/cpu-usage``
Now, image we need to add "kubernetes-io" prefix to it to distinguish system and user metrics. We will have either:
1. ``/api/v1/model/metrics/kubernetes-io/cpu-usage``
or:
2. ``/api/v1/model/metrics/kubernetes-io:cpu-usage``
In case of 1, it may be misleading whether "kubernetes-io" is part of a path or a metric name. So, together with @mwielgus we thought that 2. is better. However, we don't have a strong opinion on this, and if you think that 1. is better we can of course implement 1.

**Comment by vishh** [comment]
> Imagine, we have "cpu-usage" metric in Heapster's. If we want to query it by model API, we call:
/api/v1/model/metrics/cpu-usage

I would suggest not designing this feature based on the v1 Heapster API. 
[Here](https://github.com/kubernetes/heapster/issues/476) is an issue to update the API. 
[The new metrics APIs in heapster](https://github.com/kubernetes/heapster/blob/master/expapi/v2/api.go) are a good example of how the APIs should look like. 

One option is to not namespace supported system metrics and instead namespace just the application (aka. service) metrics (app.kubernetes.io or seomthing similar).

**Comment by bgrant0607** [comment]
Are all custom metrics going to come from heapster? Are we going to need full URLs to specify them?

Also:
`cpu-usage` and `memory-usage` are extremely imprecise.  And by "average", I assume that it's averaged across the pod metrics, which themselves are averages computed over some time period?

The proposed metrics API is:
```
// Last overlapping 10s, 1m, 1h and 1d as a start
// Updated every 10s, so the 10s window is sequential and the rest are
// rolling.
type MetricsWindows map[time.Duration]DerivedMetrics
type DerivedMetrics struct {
    Mean       ResourceUsage `json:"mean"`
    Max        ResourceUsage `json:"max"`
    NinetyFive ResourceUsage `json:"95th"`
}
```
In a world of custom metrics, I'd imagine different use cases targeting different time windows and percentiles. Even if the initial metric names are non-specific, keep in mind that we may need to introduce more specific variants in the future.

**Comment by bgrant0607** [comment]
Heapster API file:
https://github.com/kubernetes/heapster/blob/master/expapi/v2/types/metrics_types.go

**Comment by vishh** [comment]
IIRC, we decided to drop the `kubernetes.io` prefix for supported resources like `cpu` and `memory` to reduce the number of characters users have to include in their config.

**Comment by bgrant0607** [comment]
Re. DoS from creating too many pods: That's what ResourceQuota is for.

**Comment by bgrant0607** [comment]
I agree with @vishh. Omit kubernetes.io.

**Comment by jszczepkowski** [comment]
> Are all custom metrics going to come from heapster? Are we going to need full URLs to specify them?

Yes, custom metrics will come from heapster. We will not need full URL to specify them, just pod identifier and metric name.

> Also:
cpu-usage and memory-usage are extremely imprecise. And by "average", I assume that it's averaged across the pod metrics, which themselves are averages computed over some time period?

We compute the average across the latest sample for each pod.

> In a world of custom metrics, I'd imagine different use cases targeting different time windows and percentiles. Even if the initial metric names are non-specific, keep in mind that we may need to introduce more specific variants in the future.

In future, yes. This may be introduced assuming some default values for the current characteristics.

> I agree with @vishh. Omit kubernetes.io.

So, your suggestion is to have prefix only for custom metrics?

**Comment by bgrant0607** [comment]
Yes, prefix only custom metrics.

**Comment by erictune** [comment]
While it might not be a strict graduation requirement, it would be good to have user docs for HPA before 1.1 is released.  I see that there is `docs/proposals/horizontal-pod-autoscaler.md`, which is a good doc.  However, we need user-targeted documentation which gives a concrete example of how to use HPA, and which states how it works at v1.1, and doesn't talk about planned features.

**Comment by jszczepkowski** [comment]
> While it might not be a strict graduation requirement, it would be good to have user docs for HPA before 1.1 is released.

We are going to write additional user doc for HPA. In addition to the proposal, we also HPA example: ``examples/experimental/horizontal-pod-autoscaling``

**Comment by jszczepkowski** [comment]
Implemented in #15706.
